packages/protocol/contracts/rrp/AirnodeRrp.sol

```diff
@@ -264,60 +223,92 @@ contract AirnodeRrp is
     function fulfill(
         bytes32 requestId,
         address airnode,
-        uint256 statusCode,
-        bytes calldata data,
         address fulfillAddress,
-        bytes4 fulfillFunctionId
-    )
-        external
-        override
-        onlyCorrectFulfillmentParameters(
-            requestId,
-            airnode,
-            fulfillAddress,
-            fulfillFunctionId
-        )
-        returns (bool callSuccess, bytes memory callData)
-    {
+        bytes4 fulfillFunctionId,
+        bytes calldata data,
+        bytes calldata signature
+    ) external override returns (bool callSuccess, bytes memory callData) {
+        require(
+            keccak256(
+                abi.encodePacked(
+                    airnode,
+                    msg.sender,
+                    fulfillAddress,
+                    fulfillFunctionId
+                )
+            ) == requestIdToFulfillmentParameters[requestId],
+            "Invalid request fulfillment"
+        );
+        require(
+            (
+                keccak256(abi.encodePacked(requestId, data))
+                    .toEthSignedMessageHash()
+            ).recover(signature) == airnode,
+            "Invalid signature"
+        );
         delete requestIdToFulfillmentParameters[requestId];
-        emit FulfilledRequest(airnode, requestId, statusCode, data);
         (callSuccess, callData) = fulfillAddress.call( // solhint-disable-line avoid-low-level-calls
-            abi.encodeWithSelector(
-                fulfillFunctionId,
-                requestId,
-                statusCode,
-                data
-            )
+            abi.encodeWithSelector(fulfillFunctionId, requestId, data)
         );
-        require(callSuccess, "Fulfillment failed");
+        if (callSuccess) {
+            emit FulfilledRequest(airnode, requestId, data);
+        } else {
+            // We do not bubble up the revert string from `callData`
+            emit FailedRequest(
+                airnode,
+                requestId,
+                "Fulfillment failed unexpectedly"
```

Comment:
Shouldn't this be more descriptive? Like `Fulfillment callback failed unexpectedly.`?

Reply:
Probably. My reasoning was to stress that the failure is in the requester callback and not our implementation.

Reply:
Actually it should probably say `"Fulfillment reverted unexpectedly"`, then it would be obvious that the failure happened in the fulfillment transaction. I'm not a fan of using "callback" because that is not used in the protocol context to refer to fulfillments.